### PR TITLE
Feature - Import reading plan

### LINKED ISSUE #180: Easier to use method to import/use user-defined reading plans
- [ ] Validate files -- check for duplicate days, unparseable reading data

### LINKED ISSUE #570: Reading plan issues
*Sent by Errors Andbible (errors.andbible@gmail.com). Created by [fire](https://fire.fundersclub.com/).*

---
  

Hello AndBible,  

  

Two points of feedback in the date based reading plans:

  

1: I'm loving the new date based readings plans on my phone (OnePlus 5), however they are not working on my android based ereader (device info below). I've put my RobertRoberts plan in a jsword/readingplan folder (same as I did for my phone), but it's not being found by AndBible on my ereader (everything else with AndBible works fine). It would be great if you could find out why & update the implementation so that it would work on more devices.  
  
Device info:   
And Bible version: 3.2.340  
Android version: 6.0.1  
Android SDK version: ??  
Manufacturer: Onyx Boox  
Model: Boox Nova Pro  
  
2: On another note, while it was fine for me to create the folder & copy the readings plan file into it on my phone, many of my friends don't have the know-how to do this on their phones. An easier implementation for non-techsavy people would be great.

  

Another option would be to just include the Roberts reading plan by default, so that we wouldn't have to install it. I get it if you don't want to appear affiliated in any direction, but the readings plan is used by 10's of thousands around the world & some other bible apps (YouVersion for example), come with this reading plan included.  
  
I've attached the Roberts reading plan that you were using for testing, with the description updated (I've shared this version with friends on Facebook)  
  
Many thanks,

  

---
*Attachments:*

1. [Roberts.properties](https://firebot-prod-media.s3.amazonaws.com:443/email-attachments/4b12f066-5f34-4204-ad59-89ca5fa7e487/Roberts.properties)

### LINKED ISSUE #918: Crash on reading plan reset
**Describe the bug**
User reports that AB crashed when resetting reading plan. https://support.andbible.org/scp/tickets.php?id=274

**Bug was found on And Bible version**
Build 399

Have not reproduced yet. Log has no detail beside stack trace.
```
12-28 09:12:31.257 18862 18862 E AndroidRuntime: FATAL EXCEPTION: main
12-28 09:12:31.257 18862 18862 E AndroidRuntime: Process: net.bible.android.activity, PID: 18862
12-28 09:12:31.257 18862 18862 E AndroidRuntime: kotlin.UninitializedPropertyAccessException: lateinit property mReadings has not been initialized
12-28 09:12:31.257 18862 18862 E AndroidRuntime: 	at net.bible.android.view.activity.readingplan.DailyReading.access$getMReadings$p(DailyReading.kt:55)
12-28 09:12:31.257 18862 18862 E AndroidRuntime: 	at net.bible.android.view.activity.readingplan.DailyReading$onOptionsItemSelected$1.invoke(DailyReading.kt:332)
12-28 09:12:31.257 18862 18862 E AndroidRuntime: 	at net.bible.android.view.activity.readingplan.DailyReading$onOptionsItemSelected$1.invoke(DailyReading.kt:55)
12-28 09:12:31.257 18862 18862 E AndroidRuntime: 	at net.bible.android.view.activity.base.Dialogs$sam$net_bible_android_view_activity_base_Callback$0.okay(Unknown Source:2)
12-28 09:12:31.257 18862 18862 E AndroidRuntime: 	at net.bible.android.view.activity.base.Dialogs$showMsg$1$dlgBuilder$1.onClick(Dialogs.kt:104)
12-28 09:12:31.257 18862 18862 E AndroidRuntime: 	at com.android.internal.app.AlertController$ButtonHandler.handleMessage(AlertController.java:177)
12-28 09:12:31.257 18862 18862 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:107)
12-28 09:12:31.257 18862 18862 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:241)
12-28 09:12:31.257 18862 18862 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7617)
12-28 09:12:31.257 18862 18862 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
12-28 09:12:31.257 18862 18862 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
12-28 09:12:31.257 18862 18862 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:941)
```

### LINKED ISSUE #938: Read Reading Plans from addon module (similar way as fonts)

### PR DESCRIPTION
**Pull request by timbze**
Closes #180
Closes #570
Closes #938
Closes #918 (4818070d)

- [ ] Reading plan import validation for #180

### CODE REVIEW

File: app/src/main/java/net/bible/service/readingplan/ReadingPlanTextFileDao.kt [hunk 1]
```diff
@@ -243,21 +249,30 @@ class ReadingPlanTextFileDao {
...
                 }
 
                 val byteArrayForReuse = ByteArrayOutputStream().apply { write(inputStreamRaw.readBytes()) }
                 properties.load(ByteArrayInputStream(byteArrayForReuse.toByteArray()))
                 properties.planCode = planCode
                 properties.numberOfPlanDays = getNumberOfPlanDays(properties)
-                properties.versification = getReadingPlanVersification(properties)
-                properties.isDateBasedPlan = properties["1"].toString().contains("^([a-z]|[A-Z]){3}-([0-9]{1,2});".toRegex())
-                getNameAndDescFromProperties(ByteArrayInputStream(byteArrayForReuse.toByteArray()), properties)
+                properties.versification = getReadingPlanVersification(properties, userReadingPlanModule?.book?.getProperty(VERSIFICATION))
```
Comment (timbze) [inline app/src/main/java/net/bible/service/readingplan/ReadingPlanTextFileDao.kt:268]: @tuomas2 do you know a better way to get versification from Addon Book (reading plan module in this case)?

Comment (tuomas2) [reply app/src/main/java/net/bible/service/readingplan/ReadingPlanTextFileDao.kt:268]: book.versification.name works for JSword's `Book` class

Comment (timbze) [reply app/src/main/java/net/bible/service/readingplan/ReadingPlanTextFileDao.kt:268]: Hmm, it seemed that `Book` did not have any `versification` property or method. Maybe I'll have to look some more

Comment (tuomas2) [reply app/src/main/java/net/bible/service/readingplan/ReadingPlanTextFileDao.kt:268]: Ah it's in SwordBook only

**Comment by agrogers** [comment]
Where can i get reading plans from?

**Comment by timbze** [comment]
> Where can i get reading plans from?

Here is the date-based reading plan I converted to module to test this PR with.
[roberts.zip](https://github.com/AndBible/and-bible/files/6726676/roberts.zip)

You can restore it from regular documents restore and find it in Reading plans, but the "Import" option in reading plan does not work yet

**Comment by agrogers** [comment]
Thanks

**Comment by timbze** [comment]
The import validation mentioned in #180 is still needed, but don't think that blocks for this being merged.

Doing that validation is not very trivial, already spent some time working on it, and will still try to get it done, but don't think it's urgent.

**Comment by tuomas2** [comment]
I'll review this soon then!

**Comment by tuomas2** [comment]
Looks good, let's try!

**Pull request merged**